File: 146-lru-cache/lru-cache.py
from collections import deque

class LRUCache:

    def __init__(self, capacity: int):
        self.queue = deque()
        self.cache = dict()
        self.capacity = capacity

    def get(self, key: int) -> int:
        if key in self.cache:
            if key in self.queue:
                del self.queue[self.queue.index(key)]
            self.queue.append(key)

            return self.cache[key]

        return -1
        

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.queue.remove(key)
        elif len(self.cache) >= self.capacity:
            del self.cache[self.queue.popleft()]

        self.queue.append(key)
        self.cache[key] = value



# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146-lru-cache/lru-cache.cpp
struct LRUNode {
    int key;
    int value;
};

class LRUCache {
private:
    int capacity;
    std::unordered_map<int, std::list<LRUNode>::iterator> cache;
    std::list<LRUNode> order;

public:
    LRUCache(int capacity) : capacity(capacity) {
        cache.reserve(capacity);
    }

    int get(int key) {
        auto it = cache.find(key);
        if (it == cache.end())
            return -1;

        order.splice(order.begin(), order, it->second);

        return it->second->value;
    }

    void put(int key, int value) {
        auto it = cache.find(key);

        if (it != cache.end()) {
            it->second->value = value;

            order.splice(order.begin(), order, it->second);
            return;
        }

        if (order.size() == capacity) {
            auto last = order.back();
            cache.erase(last.key);
            order.pop_back();
        }

        order.push_front({key, value});
        cache[key] = order.begin();
    }
};


/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */